File: Source/AlphaTab.Test/Importer/Gp3ImporterTest.cs
﻿using AlphaTab.Model;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace AlphaTab.Test.Importer
{
    [TestClass]
    public class Gp3ImporterTest : GpImporterTestBase
    {
        [TestMethod, AsyncTestMethod]
        public void TestScoreInfo()
        {
            PrepareImporterWithFile("GuitarPro3/Test01.gp3", reader =>
            {
                var score = reader.ReadScore();

                Assert.AreEqual("Title", score.Title);
                Assert.AreEqual("Subtitle", score.SubTitle);
                Assert.AreEqual("Artist", score.Artist);
                Assert.AreEqual("Album", score.Album);
                Assert.AreEqual("Music", score.Words); // no words in gp4
                Assert.AreEqual("Music", score.Music);
                Assert.AreEqual("Copyright", score.Copyright);
                Assert.AreEqual("Tab", score.Tab);
                Assert.AreEqual("Instructions", score.Instructions);
                Assert.AreEqual("Notice1\r\nNotice2", score.Notices);
                Assert.AreEqual(5, score.MasterBars.Count);
                Assert.AreEqual(1, score.Tracks.Count);
                Assert.AreEqual("Track 1", score.Tracks[0].Name);

                Render(score);
            });
        }

        [TestMethod, AsyncTestMethod]
        public void TestNotes()
        {
            PrepareImporterWithFile("GuitarPro3/Test02.gp3", reader =>
            {
                var score = reader.ReadScore();
                CheckTest02Score(score);
                Render(score);
            });
        }

        [TestMethod, AsyncTestMethod]
        public void TestTimeSignatures()
        {
            PrepareImporterWithFile("GuitarPro3/Test03.gp3", reader =>
            {
                var score = reader.ReadScore();

                CheckTest03Score(score);
                Render(score);
            });
        }

        [TestMethod, AsyncTestMethod]
        public void TestDead()
        {
            PrepareImporterWithFile("GuitarPro3/TestDead.gp3", reader =>
            {
                var score = reader.ReadScore();
                CheckDead(score);
                Render(score);
            });
        }

        [TestMethod, AsyncTestMethod]
        public void TestAccentuation()
        {
            PrepareImporterWithFile("GuitarPro3/TestAccentuations.gp3", reader =>
            {
                var score = reader.ReadScore();

                Assert.IsTrue(score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[0].Notes[0].IsGhost);
                // it seems accentuation is handled as Forte Fortissimo
                Assert.AreEqual(DynamicValue.FFF, score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[1].Notes[0].Dynamics);
                Assert.IsTrue(score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[3].Notes[0].IsLetRing);
                Render(score);
            });
        }

        [TestMethod]
        public void TestGuitarPro3Harmonics()
        {
            // TODO: Find out about GP3 harmonics!
            //var reader = PrepareImporterWithFile("GuitarPro3/TestHarmonics.gp3");
            //var score = reader.ReadScore();

            //Assert.AreEqual(HarmonicType.Natural, score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[0].Notes[0].HarmonicType);
            //Assert.AreEqual(HarmonicType.Artificial, score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[1].Notes[0].HarmonicType);
        }

        [TestMethod, AsyncTestMethod]
        public void TestHammer()
        {
            PrepareImporterWithFile("GuitarPro3/TestHammer.gp3", reader =>
            {
                var score = reader.ReadScore();
                CheckHammer(score);
                Render(score);
            });
        }

        [TestMethod, AsyncTestMethod]
        public void TestBend()
        {
            PrepareImporterWithFile("GuitarPro3/TestBends.gp3", reader =>
            {
                var score = reader.ReadScore();
                CheckBend(score);
                Render(score);
            });
        }

        [TestMethod, AsyncTestMethod]
        public void TestSlides()
        {
            PrepareImporterWithFile("GuitarPro3/TestSlides.gp3", reader =>
            {
                var score = reader.ReadScore();

                Assert.AreEqual(SlideOutType.Shift, score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[0].GetNoteOnString(5).SlideOutType);
                Assert.AreEqual(SlideOutType.Shift, score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[2].GetNoteOnString(2).SlideOutType);
                Render(score);
            });
        }

        [TestMethod, AsyncTestMethod]
        public void TestGuitarPro3Vibrato()
        {
            // TODO: Check why this vibrato is not recognized
            PrepareImporterWithFile("GuitarPro3/TestVibrato.gp3", reader =>
            {
                var score = reader.ReadScore();
                CheckVibrato(score, false);
                Render(score);
            });
        }

        [TestMethod, AsyncTestMethod]
        public void TestOtherEffects()
        {
            PrepareImporterWithFile("GuitarPro3/TestOtherEffects.gp3", reader =>
            {
                var score = reader.ReadScore();

                Assert.IsTrue(score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[2].Tap);
                Assert.IsTrue(score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[3].Slap);

                Assert.IsTrue(score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[0].Pop);
                Assert.IsTrue(score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[1].FadeIn);

                Assert.IsTrue(score.Tracks[0].Staves[0].Bars[3].Voices[0].Beats[0].HasChord);
                Assert.AreEqual("C", score.Tracks[0].Staves[0].Bars[3].Voices[0].Beats[0].Chord.Name);
                Assert.AreEqual("Text", score.Tracks[0].Staves[0].Bars[3].Voices[0].Beats[1].Text);
                Assert.IsTrue(score.Tracks[0].Staves[0].Bars[4].Voices[0].Beats[0].GetAutomation(AutomationType.Tempo) != null);
                Assert.AreEqual(120.0, score.Tracks[0].Staves[0].Bars[4].Voices[0].Beats[0].GetAutomation(AutomationType.Tempo).Value);
                Assert.IsTrue(score.Tracks[0].Staves[0].Bars[4].Voices[0].Beats[0].GetAutomation(AutomationType.Instrument) != null);
                Assert.AreEqual(25.0, score.Tracks[0].Staves[0].Bars[4].Voices[0].Beats[0].GetAutomation(AutomationType.Instrument).Value);
                Render(score);
            });
        }

        [TestMethod, AsyncTestMethod]
        public void TestStroke()
        {
            PrepareImporterWithFile("GuitarPro3/TestStrokes.gp3", reader =>
            {
                var score = reader.ReadScore();

                Assert.AreEqual(BrushType.BrushDown, score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[0].BrushType);
                Assert.AreEqual(BrushType.BrushUp, score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[1].BrushType);
                Render(score);
            });
        }

        [TestMethod, AsyncTestMethod]
        public void TestTuplets()
        {
            PrepareImporterWithFile("GuitarPro3/TestTuplets.gp3", reader =>
            {
                var score = reader.ReadScore();
                CheckTuplets(score);
                Render(score);
            });
        }

        [TestMethod, AsyncTestMethod]
        public void TestRanges()
        {
            PrepareImporterWithFile("GuitarPro3/TestRanges.gp3", reader =>
            {
                var score = reader.ReadScore();

                Assert.IsTrue(score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[1].Notes[0].IsLetRing);
                Assert.IsTrue(score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[2].Notes[0].IsLetRing);
                Assert.IsTrue(score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[3].Notes[0].IsLetRing);
                Assert.IsTrue(score.Tracks[0].Staves[0].Bars[2].Voices[0].Beats[0].Notes[0].IsLetRing);
                Render(score);
            });
        }

        [TestMethod, AsyncTestMethod]
        public void TestEffects()
        {
            PrepareImporterWithFile("GuitarPro3/Effects.gp3", reader =>
            {
                var score = reader.ReadScore();
                CheckEffects(score);
                Render(score);
            });
        }

        [TestMethod, AsyncTestMethod]
        public void TestStrings()
        {
            PrepareImporterWithFile("GuitarPro3/TestStrings.gp3", reader =>
            {
                var score = reader.ReadScore();
                CheckStrings(score);
                Render(score);
            });
        }
    }
}

File: test/importer/Gp3Importer.test.ts
import { AutomationType } from '@src/model/Automation';
import { BrushType } from '@src/model/BrushType';
import { DynamicValue } from '@src/model/DynamicValue';
import type { Score } from '@src/model/Score';
import { SlideOutType } from '@src/model/SlideOutType';
import { GpImporterTestHelper } from '@test/importer/GpImporterTestHelper';
import { HarmonicType } from '@src/model/HarmonicType';
import { expect } from 'chai';

describe('Gp3ImporterTest', () => {
    it('score-info', async () => {
        const reader = await GpImporterTestHelper.prepareImporterWithFile('guitarpro3/score-info.gp3');
        const score: Score = reader.readScore();
        expect(score.title).to.equal('Title');
        expect(score.subTitle).to.equal('Subtitle');
        expect(score.artist).to.equal('Artist');
        expect(score.album).to.equal('Album');
        expect(score.words).to.equal('Music'); // no words in gp4

        expect(score.music).to.equal('Music');
        expect(score.copyright).to.equal('Copyright');
        expect(score.tab).to.equal('Tab');
        expect(score.instructions).to.equal('Instructions');
        expect(score.notices).to.equal('Notice1\r\nNotice2');
        expect(score.masterBars.length).to.equal(5);
        expect(score.tracks.length).to.equal(1);
        expect(score.tracks[0].name).to.equal('Track 1');
    });

    it('notes', async () => {
        const reader = await GpImporterTestHelper.prepareImporterWithFile('guitarpro3/notes.gp3');
        const score: Score = reader.readScore();
        GpImporterTestHelper.checkNotes(score);
    });

    it('time-signatures', async () => {
        const reader = await GpImporterTestHelper.prepareImporterWithFile('guitarpro3/time-signatures.gp3');
        const score: Score = reader.readScore();
        GpImporterTestHelper.checkTimeSignatures(score);
    });

    it('dead', async () => {
        const reader = await GpImporterTestHelper.prepareImporterWithFile('guitarpro3/dead.gp3');
        const score: Score = reader.readScore();
        GpImporterTestHelper.checkDead(score);
    });

    it('accentuations', async () => {
        const reader = await GpImporterTestHelper.prepareImporterWithFile('guitarpro3/accentuations.gp3');
        const score: Score = reader.readScore();
        expect(score.tracks[0].staves[0].bars[0].voices[0].beats[0].notes[0].isGhost).to.be.equal(true);
        // it seems accentuation is handled as Forte Fortissimo
        expect(score.tracks[0].staves[0].bars[0].voices[0].beats[1].notes[0].dynamics).to.equal(DynamicValue.FFF);
        expect(score.tracks[0].staves[0].bars[0].voices[0].beats[3].notes[0].isLetRing).to.be.equal(true);
    });

    it('harmonics', async () => {
        const reader = await GpImporterTestHelper.prepareImporterWithFile('guitarpro3/harmonics.gp3');
        const score: Score = reader.readScore();
        expect(score.tracks[0].staves[0].bars[0].voices[0].beats[0].notes[0].harmonicType).to.be.equal(
            HarmonicType.Natural
        );
        expect(score.tracks[0].staves[0].bars[0].voices[0].beats[1].notes[0].harmonicType).to.be.equal(
            HarmonicType.Artificial
        );
        expect(score.tracks[0].staves[0].bars[0].voices[0].beats[2].notes[0].harmonicType).to.be.equal(
            HarmonicType.Artificial
        );
        expect(score.tracks[0].staves[0].bars[0].voices[0].beats[3].notes[0].harmonicType).to.be.equal(
            HarmonicType.Artificial
        );
        expect(score.tracks[0].staves[0].bars[0].voices[0].beats[4].notes[0].harmonicType).to.be.equal(
            HarmonicType.Artificial
        );
    });

    it('hammer', async () => {
        const reader = await GpImporterTestHelper.prepareImporterWithFile('guitarpro3/hammer.gp3');
        const score: Score = reader.readScore();
        GpImporterTestHelper.checkHammer(score);
    });

    it('bends', async () => {
        const reader = await GpImporterTestHelper.prepareImporterWithFile('guitarpro3/bends.gp3');
        const score: Score = reader.readScore();
        GpImporterTestHelper.checkBend(score);
    });

    it('slides', async () => {
        const reader = await GpImporterTestHelper.prepareImporterWithFile('guitarpro3/slides.gp3');
        const score: Score = reader.readScore();
        expect(score.tracks[0].staves[0].bars[0].voices[0].beats[0].getNoteOnString(5)!.slideOutType).to.equal(
            SlideOutType.Shift
        );
        expect(score.tracks[0].staves[0].bars[0].voices[0].beats[2].getNoteOnString(2)!.slideOutType).to.equal(
            SlideOutType.Shift
        );
    });

    it('vibrato', async () => {
        // TODO: Check why this vibrato is not recognized
        const reader = await GpImporterTestHelper.prepareImporterWithFile('guitarpro3/vibrato.gp3');
        const score: Score = reader.readScore();
        GpImporterTestHelper.checkVibrato(score, false);
    });

    it('other-effects', async () => {
        const reader = await GpImporterTestHelper.prepareImporterWithFile('guitarpro3/other-effects.gp3');
        const score: Score = reader.readScore();
        expect(score.tracks[0].staves[0].bars[0].voices[0].beats[2].tap).to.be.equal(true);
        expect(score.tracks[0].staves[0].bars[0].voices[0].beats[3].slap).to.be.equal(true);
        expect(score.tracks[0].staves[0].bars[1].voices[0].beats[0].pop).to.be.equal(true);
        expect(score.tracks[0].staves[0].bars[1].voices[0].beats[1].fadeIn).to.be.equal(true);
        expect(score.tracks[0].staves[0].bars[3].voices[0].beats[0].hasChord).to.be.equal(true);
        expect(score.tracks[0].staves[0].bars[3].voices[0].beats[0].chord!.name).to.equal('C');
        expect(score.tracks[0].staves[0].bars[3].voices[0].beats[1].text).to.equal('Text');
        expect(score.tracks[0].staves[0].bars[4].voices[0].beats[0].getAutomation(AutomationType.Tempo)).to.be.ok;
        expect(
            score.tracks[0].staves[0].bars[4].voices[0].beats[0].getAutomation(AutomationType.Tempo)!.value
        ).to.equal(120);
        expect(score.tracks[0].staves[0].bars[4].voices[0].beats[0].getAutomation(AutomationType.Instrument)).to.be.ok;
        expect(
            score.tracks[0].staves[0].bars[4].voices[0].beats[0].getAutomation(AutomationType.Instrument)!.value
        ).to.equal(25);
    });

    it('strokes', async () => {
        const reader = await GpImporterTestHelper.prepareImporterWithFile('guitarpro3/strokes.gp3');
        const score: Score = reader.readScore();
        expect(score.tracks[0].staves[0].bars[0].voices[0].beats[0].brushType).to.equal(BrushType.BrushDown);
        expect(score.tracks[0].staves[0].bars[0].voices[0].beats[1].brushType).to.equal(BrushType.BrushUp);
    });

    it('tuplets', async () => {
        const reader = await GpImporterTestHelper.prepareImporterWithFile('guitarpro3/tuplets.gp3');
        const score: Score = reader.readScore();
        GpImporterTestHelper.checkTuplets(score);
    });

    it('ranges', async () => {
        const reader = await GpImporterTestHelper.prepareImporterWithFile('guitarpro3/ranges.gp3');
        const score: Score = reader.readScore();
        expect(score.tracks[0].staves[0].bars[1].voices[0].beats[1].notes[0].isLetRing).to.be.equal(true);
        expect(score.tracks[0].staves[0].bars[1].voices[0].beats[2].notes[0].isLetRing).to.be.equal(true);
        expect(score.tracks[0].staves[0].bars[1].voices[0].beats[3].notes[0].isLetRing).to.be.equal(true);
        expect(score.tracks[0].staves[0].bars[2].voices[0].beats[0].notes[0].isLetRing).to.be.equal(true);
    });

    it('effects', async () => {
        const reader = await GpImporterTestHelper.prepareImporterWithFile('guitarpro3/effects.gp3');
        const score: Score = reader.readScore();
        GpImporterTestHelper.checkEffects(score);
    });

    it('strings', async () => {
        const reader = await GpImporterTestHelper.prepareImporterWithFile('guitarpro3/strings.gp3');
        const score: Score = reader.readScore();
        GpImporterTestHelper.checkStrings(score);
    });
});
